internal/dcache/gc/gc.go

```diff
@@ -0,0 +1,259 @@
+								// There is no need for deletion of the other chunks in this RV, as the RV went offline.
+								break
+							}
+
+							//
+							// If UpdateComponentRVState() succeeds, marking this component RV as offline,
+							// we can safely carry on with the write since we are guaranteed that these
+							// chunks which we could not write to this component RV will be later sync'ed
+							// from one of the good component RVs.
+							//
+							log.Warn("GC::removeAllChunksForFile: Deletion of chunk: %d to %s/%s on node %s failed, "+
+								"marked RV offline, file: %s", chunkIdx, rv.Name, mv, targetNodeId)
+							// There is no need for deletion of the other chunks in this RV, as the RV went offline.
+							break
+						}
+
+						if rpcErr.GetCode() == models.ErrorCode_NeedToRefreshClusterMap {
+							log.Info("GC::removeAllChunksForFile: Need to refresh the cluster map, file: %s, rv: %s, err: %v",
+								file.Filename, rv.Name, rpcErr)
+
+							//
+							// We allow 5 refreshes of the clustermap for resiliency, before we fail the delete of a file.
+							// This is to allow multiple changes to the MV during the course of a deleting.
+							// It's unlikely but we need to be resilient.
+							//
+							if retryCnt > 5 {
+								log.Err("GC::removeAllChunksForFile: Max retries for updating the clusermap exhausted, "+
+									"rv: %v, file: %s: %v", rv, file.Filename, rpcErr)
+								common.Assert(false, file, rpcErr)
+								return
+							}
+
+							//
+							// Retry till the next epoch, ensuring that the clustermap is refreshed from what we
+							// have cached right now.
+							//
+							errCM := cm.RefreshClusterMap(lastClusterMapEpoch)
+							if errCM != nil {
+								log.Err("GC::removeAllChunksForFile: Failed to refresh the cluster map, rv: %s, file: %s: %v",
+									rv.Name, file.Filename, errCM)
+								common.Assert(false, file, errCM)
+								return
+							}
+
+							retryCnt++
+							goto retry
+						}
+
+						common.Assert(rpcErr.GetCode() == models.ErrorCode_ChunkNotFound &&
+							rv.State == string(dcache.StateSyncing), file.Filename, rv, rvs, rpcErr, rpcResp)
+					}
+
+					// Update the progress of deletion for the RV.
+					deleteProgressForRvs[rv.Name] = getNextChunkIdxInMV(chunkIdx, numMvs)
+				}
+			}
+		}
+	}
+
+	// After removing all the chunks from all the rvs, we can remove the file layout.
+	deletedFile := dcache.GetDeletedFileName(file.Filename, file.FileID)
+	log.Debug("GC::removeAllChunksForFile: removing file layout for file: %s", deletedFile)
+
+	err := metadata_manager.DeleteFile(deletedFile)
+	if err != nil {
+		log.Err("GC::removeAllChunksForFile: failed to remove file layout for file: %s: %v", deletedFile, err)
+		common.Assert(false, file, err)
+		return
+	}
+
+}
+
+func AsyncFileChunkGarbageCollector(file *dcache.FileMetadata) {
+	if file.Size != 0 {
+		gc.deletedFileQueue <- file
+	}
+}
```

Comment:
while this will be the common way to delete chunks for a file, we should also support the case where the node that deletes a file dies, so it cannot gc the file chunks using this mechanism.
We should periodically check for .deleting files and try to GC them. We could have used that as the only mechanism, but it might have overhead of enumerating potentially large number of files, so let's support both for now.